Bash gcc vs clang mention

Noticed that both gcc and clang are referenced [together](https://github.com/Michael-F-Bryan/rust-ffi-guide/blame/cac3ba7eb8144e6599796741ebbdfa796d760831/src/introduction/index.md#L112:L115)

Is the book supposed to use GCC or clang?

How did you get file paths and line numbers in [clang's error output](https://github.com/Michael-F-Bryan/rust-ffi-guide/blame/cac3ba7eb8144e6599796741ebbdfa796d760831/src/arrays/index.md#L121-L129)?
All I get are lots of tiny object files.
```
(.text._ZN3std10sys_common12thread_local9StaticKey9lazy_init17h3cc2c767095c0390E+0x5c): undefined reference to `pthread_key_create'
libaverages.a(std-e22d8ea13fc64833.2.o): In function `std::sys_common::thread_local::StaticKey::lazy_init':
(.text._ZN3std10sys_common12thread_local9StaticKey9lazy_init17h3cc2c767095c0390E+0x75): undefined reference to `pthread_key_delete'
libaverages.a(std-e22d8ea13fc64833.2.o): In function `std::sys_common::thread_local::StaticKey::lazy_init':
```

Other than that, the output between gcc and clang only differ by the last line
**clang**
```
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
```
**gcc**
```
collect2: error: ld returned 1 exit status
```

> How did you get file paths and line numbers in clang's error output?

I'm not sure to be honest. Both gcc and clang give the same output and I didn't do anything special. Here's everything I did to reproduce it.

```bash
$ cd src/arrays/
$ rustc --crate-type staticlib averages.rs
(skipping the output about static dependencies)
$ clang -o main main.c libaverages.a
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::mutex::{{impl}}::init':
/checkout/src/libstd/sys/unix/mutex.rs:56: undefined reference to `pthread_mutexattr_init'
/checkout/src/libstd/sys/unix/mutex.rs:58: undefined reference to `pthread_mutexattr_settype'
/checkout/src/libstd/sys/unix/mutex.rs:62: undefined reference to `pthread_mutexattr_destroy'
/checkout/src/libstd/sys/unix/mutex.rs:56: undefined reference to `pthread_mutexattr_init'
/checkout/src/libstd/sys/unix/mutex.rs:58: undefined reference to `pthread_mutexattr_settype'
/checkout/src/libstd/sys/unix/mutex.rs:62: undefined reference to `pthread_mutexattr_destroy'
/checkout/src/libstd/sys/unix/mutex.rs:110: undefined reference to `pthread_mutexattr_init'
/checkout/src/libstd/sys/unix/mutex.rs:112: undefined reference to `pthread_mutexattr_settype'
/checkout/src/libstd/sys/unix/mutex.rs:117: undefined reference to `pthread_mutexattr_destroy'
/checkout/src/libstd/sys/unix/mutex.rs:110: undefined reference to `pthread_mutexattr_init'
/checkout/src/libstd/sys/unix/mutex.rs:112: undefined reference to `pthread_mutexattr_settype'
/checkout/src/libstd/sys/unix/mutex.rs:117: undefined reference to `pthread_mutexattr_destroy'
/checkout/src/libstd/sys/unix/mutex.rs:56: undefined reference to `pthread_mutexattr_init'
/checkout/src/libstd/sys/unix/mutex.rs:58: undefined reference to `pthread_mutexattr_settype'
/checkout/src/libstd/sys/unix/mutex.rs:62: undefined reference to `pthread_mutexattr_destroy'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::create':
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::destroy':
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::create':
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::destroy':
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::get':
/checkout/src/libstd/sys/unix/thread_local.rs:33: undefined reference to `pthread_getspecific'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::create':
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::destroy':
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::set':
/checkout/src/libstd/sys/unix/thread_local.rs:27: undefined reference to `pthread_setspecific'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::create':
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::destroy':
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::get':
/checkout/src/libstd/sys/unix/thread_local.rs:33: undefined reference to `pthread_getspecific'
/checkout/src/libstd/sys/unix/thread_local.rs:33: undefined reference to `pthread_getspecific'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::create':
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
/checkout/src/libstd/sys/unix/thread_local.rs:21: undefined reference to `pthread_key_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::destroy':
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
/checkout/src/libstd/sys/unix/thread_local.rs:38: undefined reference to `pthread_key_delete'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread_local::set':
/checkout/src/libstd/sys/unix/thread_local.rs:27: undefined reference to `pthread_setspecific'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::weak::fetch':
/checkout/src/libstd/sys/unix/weak.rs:78: undefined reference to `dlsym'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::condvar::{{impl}}::init':
/checkout/src/libstd/sys/unix/condvar.rs:50: undefined reference to `pthread_condattr_setclock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::process::process_inner::{{impl}}::do_exec':
/checkout/src/libstd/sys/unix/process/process_unix.rs:206: undefined reference to `pthread_sigmask'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::pthread_attr_setstacksize':
/checkout/src/libstd/sys/unix/thread.rs:37: undefined reference to `pthread_attr_setstacksize'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::{{impl}}::new':
/checkout/src/libstd/sys/unix/thread.rs:67: undefined reference to `pthread_attr_setstacksize'
/checkout/src/libstd/sys/unix/thread.rs:72: undefined reference to `pthread_create'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::{{impl}}::join':
/checkout/src/libstd/sys/unix/thread.rs:169: undefined reference to `pthread_join'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::{{impl}}::drop':
/checkout/src/libstd/sys/unix/thread.rs:186: undefined reference to `pthread_detach'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::guard::current':
/checkout/src/libstd/sys/unix/thread.rs:338: undefined reference to `pthread_getattr_np'
/checkout/src/libstd/sys/unix/thread.rs:341: undefined reference to `pthread_attr_getguardsize'
/checkout/src/libstd/sys/unix/thread.rs:347: undefined reference to `pthread_attr_getstack'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::write':
/checkout/src/libstd/sys/unix/rwlock.rs:82: undefined reference to `pthread_rwlock_wrlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::raw_unlock':
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::write':
/checkout/src/libstd/sys/unix/rwlock.rs:82: undefined reference to `pthread_rwlock_wrlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::raw_unlock':
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::read':
/checkout/src/libstd/sys/unix/rwlock.rs:34: undefined reference to `pthread_rwlock_rdlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::rwlock::{{impl}}::raw_unlock':
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
/checkout/src/libstd/sys/unix/rwlock.rs:113: undefined reference to `pthread_rwlock_unlock'
libaverages.a(std-b98703028eeb0b98.0.o): In function `std::sys::imp::thread::guard::get_stack_start':
/checkout/src/libstd/sys/unix/thread.rs:236: undefined reference to `pthread_getattr_np'
/checkout/src/libstd/sys/unix/thread.rs:240: undefined reference to `pthread_attr_getstack'
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
$ clang --version 
clang version 4.0.1 (tags/RELEASE_401/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/sbin
```

If it makes a difference, I'm running this on my Arch laptop.

--- a/src/arrays/index.md
+++ b/src/arrays/index.md
@@ -117,7 +117,7 @@ int main() {
 And now let's compile everything.
 
 ```bash
-$ gcc main.c libaverages.a
+$ clang main.c libaverages.a
 libaverages.a(std-9a66b6a343d52844.0.o): In function `std::sys::imp::mutex::{{impl}}::init':
 /checkout/src/libstd/sys/unix/mutex.rs:56: undefined reference to `pthread_mutexattr_init'
 /checkout/src/libstd/sys/unix/mutex.rs:58: undefined reference to `pthread_mutexattr_settype'

--- a/src/bindgen/index.md
+++ b/src/bindgen/index.md
@@ -20,6 +20,8 @@ use Rust in a C/C++ codebase in the first place!
 
 ## Getting Set Up
 
+> **Note:** Bindgen **requires** clang. Installation instructions can be found [here](https://github.com/rust-lang-nursery/rust-bindgen/blob/master/book/src/requirements.md#installing-clang-39)
+
 We'll be building on top of the [great tutorial][tut] linked earlier to write
 an idiomatic Rust wrapper around `bzip2`. This assumes you've already read
 through that tutorial and will focus more on the next step, writing the actual

--- a/src/introduction/Makefile
+++ b/src/introduction/Makefile
@@ -4,7 +4,7 @@ all: libhello.so main
 
 
 libhello.so: hello.c
-	gcc -shared -fPIC -o libhello.so hello.c
+	clang -shared -fPIC -o libhello.so hello.c
 
 main: libhello.so main.rs
 	rustc -l hello -L . main.rs

--- a/src/introduction/index.md
+++ b/src/introduction/index.md
@@ -109,7 +109,7 @@ external `say_hello` symbol will be resolved at *load time*.
 > [this Stack Overflow question][static-vs-dynamic].
 
 ```bash
-$ gcc -shared -fPIC -o libhello.so hello.c
+$ clang -shared -fPIC -o libhello.so hello.c
 ```
 
 > **Note:** clang and gcc are interchangeable when compiling.

--- a/src/introduction/index.md
+++ b/src/introduction/index.md
@@ -112,6 +112,8 @@ external `say_hello` symbol will be resolved at *load time*.
 $ gcc -shared -fPIC -o libhello.so hello.c
 ```
 
+> **Note:** clang and gcc are interchangeable when compiling.
+
 The `-shared` flag tells clang to compile as a dynamically linked library
 (typically "\*.so" on Linux or "\*.dll" on Windows). You'll also need the
 `-fPIC` flag to tell the compiler to generate [*Position Independent

